api/nodes/serializers.py

```diff
@@ -124,9 +128,124 @@ def update(self, instance, validated_data):
+        node.add_contributor(contributor=contributor, auth=auth, visible=bibliographic, permissions=permissions, save=True)
+        contributor.permission = node.get_permissions(contributor)[-1]
+        contributor.bibliographic = node.get_visible(contributor)
+        contributor.node_id = node._id
+        return contributor
+
+    @staticmethod
+    def get_permissions_list(permission):
+        if permission == osf_permissions.ADMIN:
+            return [osf_permissions.READ, osf_permissions.WRITE, osf_permissions.ADMIN]
+        elif permission == osf_permissions.WRITE:
+            return [osf_permissions.READ, osf_permissions.WRITE]
+        elif permission == osf_permissions.READ:
+            return [osf_permissions.READ]
+        else:
+            return osf_permissions.DEFAULT_CONTRIBUTOR_PERMISSIONS
+
+
+class NodeContributorDetailSerializer(NodeContributorsSerializer):
+    """ Overrides node contributor serializer to make id read only and add additional methods
+    """
+    id = ser.CharField(read_only=True, source='_id', label='ID')
+
+    def update(self, instance, validated_data):
+        contributor = instance
+        auth = Auth(self.context['request'].user)
+        node = self.context['view'].get_node()
+        if 'permission' in validated_data and validated_data['permission'] != '':
+            self.set_contributor_permissions(node, validated_data['permission'], contributor, auth)
+        try:
+            node.set_visible(contributor, validated_data['bibliographic'], save=True)
+        except ValueError as e:
+            raise exceptions.ValidationError(e)
+        contributor.permission = node.get_permissions(contributor)[-1]
+        contributor.bibliographic = node.get_visible(contributor)
+        contributor.node_id = node._id
+        return contributor
+
+    def set_contributor_permissions(self, node, permission, contributor, auth):
+        """ Set contributor permissions. Checks to make sure unique admin is
+        removing own admin privilege. Based on the ContributorDetailPermissions
+        permissions class, the current user must be an admin.
+        """
+        assert node.has_permission(auth.user, osf_permissions.ADMIN), "Only admins can modify contributor permissions"
+        permissions = self.get_permissions_list(permission)
+        if not self.has_multiple_admins(node):
+            # has only one admin
+            admin = self._get_admins(node)[0]
+            if admin == contributor and osf_permissions.ADMIN not in permissions:
+                raise exceptions.ValidationError('{} is the only admin.'.format(contributor.username))
+        node.set_permissions(contributor, permissions=permissions, save=True)
+
+    def _get_admins(self, node):
+        return [
+            contributor for contributor in node.contributors
+            if node.has_permission(contributor, osf_permissions.ADMIN)
+        ]
+
+    # created due to issues node.admin_contributor_ids not working
```

Comment:
FWIW, it's not that `admin_contributor_ids` doesn't work; it's just that it returns the ids of admins on the node *and* the parents.